MauroDataMapper-NHSD/mdm-plugin-nhs-data-dictionary
# Change Paper - content updates

Collection of changes to support missing content in the Change Papers.

# Testing

I've tried to include some automated tests to verify the HTML/DITA output, but manual testing is slow and laborious...

1. Publish `mdm-plugin-nhs-data-dictionary` to Maven local and run in `mdm-application-build`
2. Run the Orchestrator and sign in
3. Click on the "Changes" nav link
4. Select an in-flight branch, then click "Run". This will give a preview of the change paper content.
5. Next click on "Branches" nav link.
6. Select a in-flight branch, click "Publish", then one of the generate buttons to produce DITA
7. Wait...
8. Download the zip file result, extract to a local folder
9. Use OxygenXML on the DITA content to produce a PDF

# Publication Date Format

Fixes #66 

Fixed the publication date to use "dd MMMM yyyy" format, as requested in NHSDigital/DataDictionaryPublication#498

![image](https://github.com/user-attachments/assets/91a8bc6f-3f11-4a97-be5b-81a744cf147e)

# Missing descriptions for "Updated descriptions"

Fixes #72 

This didn't seem to work on TEST but does on my machine (I don't understand why...), so considering this resolved apart from styling, which is NHSE responsibility:

![image](https://github.com/user-attachments/assets/5c4e2a30-0f25-4f4a-8a6b-5fab427b8155)

# Include "Change attributes" list for Classes

Fixes #75 

Classes did not appear in the list of changes because only descriptions were checked. I've updated this to perform a check on attributes within a class that might have been added/removed:

![image](https://github.com/user-attachments/assets/9f9620ff-ee1f-4524-9912-48c64bd147c2)

# Include missing tables for Attributes

Fixes #74 

Attributes had certain information missing, which I have now added.

Aliases table:
![image](https://github.com/user-attachments/assets/f34b6956-dfa9-43e8-97ea-a005d081f218)

National codes table:
![image](https://github.com/user-attachments/assets/f8655732-9af0-4b08-8b3a-d8a31eedbce4)

# Include "Format / Length" for Data Elements

![image](https://github.com/user-attachments/assets/0b09ae6a-de8c-48c5-a8f3-cc98146e122b)

# "Summary of Changes" incorrect

Resolves #71 

I think this is down to the data ingested and being modified on the server. Also, some of the supposed changes were down to missing content (above), so I think this issue is actually a red herring.

# Data Set definition changes

Resolves #76 

When including data set definitions:

1. Refer to the definitions as "Changed Data Set" instead of "Unchanged Items", like the original Borland Together change papers.
2. Don't include Data Elements and Attributes related to data sets multiple times - if a Data Element or Attribute was changed, it does not need to be added a second time, since this breaks the unique topic keys of the DITA output.

# Exclude diffs containing HTML tables

Fixes #77 

There is an issue with `DaisyDiffHelper` when comparing HTML content containing tables, the added/removed tags/markup that it adds will intermingle with the table markup, causing a table structure that is not accurate. This is especially troublesome when converting the HTML into DITA because that is much stricter on the XML schema. Temporary solution is to not perform a diff on such content - just show the current content with a message to explain.

# Include data set specifications

Resolves #73 

Added the missing data set specification tables to the change paper (preview and PDF):

![image](https://github.com/user-attachments/assets/d59b56d8-503e-4353-938f-14ffb83435f9)
